include/pybind11/functional.h

```diff
@@ -15,6 +15,47 @@
 
 PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
 PYBIND11_NAMESPACE_BEGIN(detail)
+PYBIND11_NAMESPACE_BEGIN(type_caster_std_function_specializations)
+
+// ensure GIL is held during functor destruction
+struct func_handle {
+    function f;
+#if !(defined(_MSC_VER) && _MSC_VER == 1916 && defined(PYBIND11_CPP17))
+    // This triggers a syntax error under very special conditions (very weird indeed).
+    explicit
+#endif
+        func_handle(function &&f_) noexcept
+        : f(std::move(f_)) {
+    }
+    func_handle(const func_handle &f_) { operator=(f_); }
+    func_handle &operator=(const func_handle &f_) {
+        gil_scoped_acquire acq;
+        f = f_.f;
+        return *this;
+    }
+    ~func_handle() {
+        gil_scoped_acquire acq;
+        function kill_f(std::move(f));
+    }
+};
+
+// to emulate 'move initialization capture' in C++11
+struct func_wrapper_base {
+    func_handle hfunc;
+    explicit func_wrapper_base(func_handle &&hf) noexcept : hfunc(std::move(hf)) {}
+};
+
+template <typename Return, typename... Args>
+struct func_wrapper : func_wrapper_base {
+    using func_wrapper_base::func_wrapper_base;
+    Return operator()(Args... args) const {
```

Comment:
Could you please try if this works?

```
template <typename Return>
struct func_wrapper : func_wrapper_base {
    using func_wrapper_base::func_wrapper_base;
    template <typename... Args>
    Return operator()(Args... args) const {
```

(and the corresponding change below)?

Reply:
No it fails with error 

```
error: no viable overloaded '='
        value = type_caster_std_function_specializations::func_wrapper<Return>(
```

The `value` has type `std::function<Return(Args...)>;`.

Reply:
Oh sorry, I wasn't thinking this through. Thanks for trying it out.